### PR TITLE
Refactor `UInt{8,16,64,128}` into one struct `UInt`

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

closes: #116 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by mmagician — COMMENTED** [review]
Looks good so far.
I wonder whether we could completely get rid of `uint8` module?
To preserve+ the functionality, it would require the `UInt<const N: usize, T: PrimInt + Debug, F: Field>` struct to implement `new_input_vec`, which would in turn require an extra trait bound `T: ToConstraintField`. These seem like natural things to have for `u16` etc. anyway.

This could easily be a follow-up PR though, but if you agree then let's open an issue for tracking.

**Comment by Pratyush** [comment]
Unfortunately I think there's a special impl of `ToConstraintField` for `[u8]`s that's used in the `new_input_vec` impl. Generalizing it to support larger sizes shouldn't be difficult, but would require moving away from calling `ToConstraintField`. Maybe that's fine though.

**Comment by mmagician** [comment]
Rather than generalizing we could just implement `ToConstraintField` for `&[u16]` etc no?

File: src/bits/uint/mod.rs [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+            output_vec.push(Self::new_witness(cs.clone(), || byte.get())?);
+        }
+        Ok(output_vec)
+    }
+
+    /// Perform modular addition of `operands`.
+    ///
+    /// The user must ensure that overflow does not occur.
+    #[tracing::instrument(target = "r1cs", skip(operands))]
+    pub fn add_many(operands: &[Self]) -> Result<Self, SynthesisError>
```
Comment (rozbb) [inline src/bits/uint/mod.rs:132]: Using this fork right now for code that previously worked. I'm getting the error `'attempt to multiply with overflow'` stemming from somewhere in this method. It doesn't say where 🤔

Comment (Pratyush) [reply src/bits/uint/mod.rs]: Fixed now!

**Review by z-tech — APPROVED** [review]

**Comment by autquis** [comment]
@Pratyush After moving `ToBytesGadget` to `convert`, it is no longer part of `prelude`. Is it intended?
I'm fixing crypto-primitive for this change. For example, in:
https://github.com/arkworks-rs/crypto-primitives/blob/8bbb4824ede5f2cf115f1a37e08410b57944a4f6/src/signature/constraints.rs#L10

It was previously enough to `use ark-r1cs-std::prelude::*` but now I need either use it explicitly or update the `prelude`.

**Pull request merged**